File: src/run/orchestrator.py
from __future__ import annotations

from pathlib import Path

from run.approval import PendingHumanApprovalGateway
from run.execution import DelegatingExecutor, DryRunExecutor
from run.interfaces import (
    AgentExecutor,
    AgentRouter,
    ApprovalGateway,
    GitWorkflow,
    InitHooks,
    MemoryStore,
    PlanStore,
    SecurityPolicy,
    VerificationRunner,
)
from run.ledger import RunLedger
from run.routing import LogicalAgentRouter
from run.types import (
    ApprovalStatus,
    ExecutionMode,
    ExecutionResult,
    PrDraft,
    RunRequest,
    RunResult,
    RunState,
    VerificationResult,
)


class NoopInitHooks:
    def before_run(self, request: RunRequest) -> None:
        pass

    def after_run(self, request: RunRequest) -> None:
        pass


class NoopSecurityPolicy:
    def assert_plan_can_run(self, plan) -> None:
        if not plan.approved:
            raise ValueError(f"Plan {plan.plan_id} is not approved.")


class NoopMemoryStore:
    def record_run(self, request: RunRequest, result: tuple[ExecutionResult, ...]) -> None:
        pass


class NoopVerificationRunner:
    def run(self, repository: Path) -> tuple[VerificationResult, ...]:
        return ()


class RunOrchestrator:
    def __init__(
        self,
        plan_store: PlanStore,
        git_workflow: GitWorkflow,
        ledger: RunLedger,
        security_policy: SecurityPolicy | None = None,
        approval_gateway: ApprovalGateway | None = None,
        router: AgentRouter | None = None,
        executor: AgentExecutor | None = None,
        verification_runner: VerificationRunner | None = None,
        memory_store: MemoryStore | None = None,
        init_hooks: InitHooks | None = None,
    ) -> None:
        self.plan_store = plan_store
        self.git_workflow = git_workflow
        self.ledger = ledger
        self.security_policy = security_policy or NoopSecurityPolicy()
        self.approval_gateway = approval_gateway or PendingHumanApprovalGateway()
        self.router = router or LogicalAgentRouter()
        self.executor = executor
        self.verification_runner = verification_runner or NoopVerificationRunner()
        self.memory_store = memory_store or NoopMemoryStore()
        self.init_hooks = init_hooks or NoopInitHooks()

    def run(self, request: RunRequest) -> RunResult:
        plan = None
        approval = None
        assignments = ()
        execution_results = ()
        verification_results = ()
        pr_draft: PrDraft | None = None

        try:
            self.ledger.record("run.created", {"plan_id": request.plan_id, "mode": request.mode.value})
            self.init_hooks.before_run(request)

            plan = self.plan_store.load_approved_plan(request.plan_id)
            self.ledger.record("plan.loaded", {"plan_id": plan.plan_id, "approved": plan.approved})

            self.security_policy.assert_plan_can_run(plan)
            self.ledger.record("security.checked", {"plan_id": plan.plan_id})

            assignments = tuple(self.router.route(request, plan))
            self.ledger.record("agents.routed", {"count": len(assignments)})

            approval = self.approval_gateway.request_execution_approval(request, plan, assignments)
            self.ledger.record("approval.requested", {"status": approval.status.value})

            if approval.status is not ApprovalStatus.APPROVED:
                state = RunState.REJECTED if approval.status is ApprovalStatus.REJECTED else RunState.APPROVAL_REQUIRED
                self.ledger.record("run.paused", {"state": state.value})
                return RunResult(
                    state=state,
                    plan=plan,
                    approval=approval,
                    assignments=assignments,
                    execution_results=execution_results,
                    verification_results=verification_results,
                    pr_draft=pr_draft,
                    ledger_path=self.ledger.path,
                )

            branch_name = self.git_workflow.prepare_branch(request)
            self.ledger.record("git.branch.prepared", {"branch": branch_name})

            executor = self.executor or self._default_executor(request)
            execution_results = tuple(executor.execute(request, plan, assignments))
            event_type = "execution.skipped" if request.mode is ExecutionMode.DRY_RUN else "execution.completed"
            self.ledger.record(event_type, {"count": len(execution_results)})

            verification_results = tuple(self.verification_runner.run(request.repository))
            self.ledger.record(
                "verification.completed",
                {"passed": all(result.passed for result in verification_results), "count": len(verification_results)},
            )

            self.memory_store.record_run(request, execution_results)
            self.ledger.record("memory.updated", {"plan_id": plan.plan_id})

            pr_draft = self.git_workflow.prepare_pr_draft(request, plan, verification_results)
            self.ledger.record("pr.prepared", {"provider": pr_draft.provider, "branch": pr_draft.branch_name})

            self.init_hooks.after_run(request)
            self.ledger.record("run.completed", {"plan_id": plan.plan_id})
            return RunResult(
                state=RunState.COMPLETED,
                plan=plan,
                approval=approval,
                assignments=assignments,
                execution_results=execution_results,
                verification_results=verification_results,
                pr_draft=pr_draft,
                ledger_path=self.ledger.path,
            )
        except Exception as exc:
            self.ledger.record("run.failed", {"error": str(exc)})
            return RunResult(
                state=RunState.FAILED,
                plan=plan,
                approval=approval,
                assignments=assignments,
                execution_results=execution_results,
                verification_results=verification_results,
                pr_draft=pr_draft,
                ledger_path=self.ledger.path,
            )

    def _default_executor(self, request: RunRequest) -> AgentExecutor:
        if request.mode is ExecutionMode.DRY_RUN:
            return DryRunExecutor()
        return DelegatingExecutor()

File: src/run/orchestrator.ts
import { resolve } from "node:path";
import type { Env } from "../fs/layout.js";
import { createExecutor, type AgentExecutor } from "./executor.js";
import { preparePrMetadata } from "./git.js";
import { RunLedger } from "./ledger.js";
import { recordRunMemory } from "./memory.js";
import { loadRunPlan, uniqueAgents } from "./plan.js";
import { evaluateRunSecurity, securityAllowsSideEffects } from "./security.js";
import { runVerification } from "./verification.js";
import type {
  AgentAssignment,
  ApprovalDecision,
  ApprovalSource,
  PrProvider,
  RunFinalState,
  RunMode,
  RunPlan,
  RunSummary
} from "./types.js";

export interface RunOrchestratorOptions {
  readonly cwd: string;
  readonly env: Env;
  readonly planPath?: string;
  readonly resumeRunId?: string;
  readonly approved: boolean;
  readonly approvalSource: ApprovalSource;
  readonly mode: RunMode;
  readonly provider?: PrProvider;
  readonly fakeAgent: boolean;
  readonly executor?: AgentExecutor;
}

const terminalStates: readonly RunFinalState[] = ["failed", "verification_failed", "completed", "needs_review"];

export async function runFromPlan(options: RunOrchestratorOptions): Promise<RunSummary> {
  const ledger = RunLedger.create(options.cwd, options.env, options.resumeRunId);
  await ledger.initialize();
  const checkpoint = options.resumeRunId ? await ledger.readCheckpoint() : undefined;

  if (checkpoint && terminalStates.includes(checkpoint.state)) {
    const plan = await loadRunPlan(checkpoint.planPath);
    const approval = approvalDecision(plan, options.approved, options.approvalSource);
    await ledger.record("run.resume.noop", { state: checkpoint.state }, checkpoint.state);
    return emptySummary(ledger, plan, approval, checkpoint.state);
  }

  const requestedPlanPath = checkpoint?.planPath ?? options.planPath;
  if (!requestedPlanPath) {
    throw new Error("apolo run requires --from-plan <path> or --resume <run-id>.");
  }
  const planPath = resolve(requestedPlanPath);

  await ledger.record(
    checkpoint ? "run.resumed" : "run.created",
    { planPath, mode: options.mode, resumeRunId: options.resumeRunId ?? null },
    "running"
  );
  const plan = await loadRunPlan(planPath);
  await ledger.record("plan.loaded", { planId: plan.id, approved: plan.approved, approvalRequired: plan.approvalRequired });

  const approval = approvalDecision(plan, options.approved, options.approvalSource);
  const security = evaluateRunSecurity(plan, approval);
  await ledger.record("security.checked", {
    decisions: security.map((decision) => ({
      gate: decision.gate,
      allowed: decision.allowed,
      severity: decision.severity,
      reason: decision.reason
    }))
  });

  if (!securityAllowsSideEffects(security)) {
    const onlyApprovalRequired = security.every((decision) => decision.allowed || decision.severity === "approval");
    const state: RunFinalState = onlyApprovalRequired ? "approval_required" : "failed";
    await ledger.record("run.paused", { state }, state);
    await ledger.checkpoint(plan.sourcePath, state);
    return emptySummary(ledger, plan, approval, state);
  }

  const assignments = routeAgents(plan);
  await ledger.record("agents.routed", {
    count: assignments.length,
    agents: assignments.map((assignment) => assignment.agent)
  });
  await ledger.record("execution.started", { mode: options.mode }, "running");

  const executor = options.executor ?? createExecutor(options.mode, options.fakeAgent);
  const executionResults = await executor.execute(plan, assignments, options.cwd);
  await ledger.record("execution.completed", {
    results: executionResults.map((result) => ({
      agent: result.agent,
      stepId: result.stepId,
      status: result.status,
      exitCode: result.exitCode ?? null
    }))
  });

  const verificationResults = await runVerification(options.cwd, plan);
  await ledger.record("verification.completed", {
    passed: verificationResults.every((result) => result.passed),
    count: verificationResults.length
  });

  await recordRunMemory(options.cwd, options.env, plan, executionResults);
  await ledger.record("memory.updated", { planId: plan.id });

  const prMetadata = await preparePrMetadata({
    cwd: options.cwd,
    runDir: ledger.runDir,
    plan,
    provider: options.provider ?? plan.provider,
    executionResults,
    verificationResults
  });
  await ledger.record("pr.prepared", {
    provider: prMetadata.provider,
    branchName: prMetadata.branchName,
    pushPolicy: prMetadata.pushPolicy
  });

  const state = finalState(executionResults, verificationResults);
  await ledger.record("run.finished", { state }, state);
  await ledger.checkpoint(plan.sourcePath, state);

  return {
    runId: ledger.runId,
    state,
    planId: plan.id,
    planPath: plan.sourcePath,
    ledgerPath: ledger.ledgerPath,
    runDir: ledger.runDir,
    approval,
    assignments,
    executionResults,
    verificationResults,
    prMetadata
  };
}

function approvalDecision(plan: RunPlan, approved: boolean, source: ApprovalSource): ApprovalDecision {
  if (plan.approved) {
    return {
      approved: true,
      source: "plan",
      reason: "Plan artifact is already approved."
    };
  }
  if (approved) {
    return {
      approved: true,
      source,
      reason: "Runtime approval was explicitly provided."
    };
  }
  return {
    approved: false,
    source: "none",
    reason: "No plan or runtime approval was provided."
  };
}

function routeAgents(plan: RunPlan): readonly AgentAssignment[] {
  return uniqueAgents(plan.steps).map((agent) => ({
    agent,
    steps: plan.steps.filter((step) => step.agent === agent)
  }));
}

function finalState(
  executionResults: readonly { readonly status: string }[],
  verificationResults: readonly { readonly passed: boolean }[]
): RunFinalState {
  if (executionResults.some((result) => result.status === "failed")) {
    return "failed";
  }
  if (executionResults.some((result) => result.status === "needs_review")) {
    return "needs_review";
  }
  if (verificationResults.some((result) => !result.passed)) {
    return "verification_failed";
  }
  return "completed";
}

function emptySummary(ledger: RunLedger, plan: RunPlan, approval: ApprovalDecision, state: RunFinalState): RunSummary {
  return {
    runId: ledger.runId,
    state,
    planId: plan.id,
    planPath: plan.sourcePath,
    ledgerPath: ledger.ledgerPath,
    runDir: ledger.runDir,
    approval,
    assignments: [],
    executionResults: [],
    verificationResults: []
  };
}
